cpu-o3: add intel topdown

tastynoob opened the pull request.

tastynoob committed:
cpu-o3: add mem stall topdown

Change-Id: Ice57ee1cc06d16cfd0510af2c32eeda811f23ebc

jensen-yan committed:
cpu-o3: add frontend topdown counter

Change-Id: Ieb276954105f40a33b52f3e957d52778d655abd9

tastynoob committed:
cpu-o3: modify load memstall, add store memstall

Change-Id: Ibe07f7180567689680fd6878163403bfa0178a3a

seanwang2046 commented on src/cpu/o3/commit.cc:
```diff
@@ -1164,8 +1165,15 @@ Commit::commitInsts()
                 cpu->perfCCT->commitMeta(head_inst->seqNum);
-                lastCommitCycle = cpu->curCycle();
                 head_inst->printDisassemblyAndResult(cpu->name());
+                if (ismispred) {
```
why let " if(ismisped)..." executed before setting the ismispred =true?

tastynoob replied:
The mispred branch still will commit

seanwang2046 commented on src/cpu/o3/fetch.cc:
```diff
@@ -245,7 +245,23 @@ Fetch::FetchStatGroup::FetchStatGroup(CPU *cpu, Fetch *fetch)
                     statistics::units::Count, statistics::units::Cycle>::get(),
              "Number of decode stalls per cycle",
-             decodeStalls / cpu->baseStats.numCycles)
+             decodeStalls / cpu->baseStats.numCycles),
```
does the "front end bound/frontend latency bound/frontend BW band" calculated every cycle or  calculated just once after the running? if calculated every cycle it may have bad impact on simulation speed. so I suggest we count the topdown PMU counters each cycle while calculate the topdown bounds just once after the running

seanwang2046 commented on src/cpu/o3/fetch.cc:
```diff
@@ -1294,7 +1323,20 @@ Fetch::tick()
+    // For N-wide machine, if frontend supplies 0 instructions:
+    // - fetchBubbles += N (count total empty slots)
+    // - fetchBubbles_max += 1 (count occurrence of all slots being empty)
+    if (!stalls[*tid_itr].decode) { // backend not stalled
```
does fetch bubble also count when flush bubble occurs? we expect that fetch bubble should not count when flush occurs, because flush penalty should be count as part of bad spec

seanwang2046 commented on src/cpu/o3/issue_queue.cc:
```diff
@@ -764,6 +781,18 @@ Scheduler::issueAndSelect()
     for (auto it : issueQues) {
         it->issueToFu();
     }
+    if (instsToFu.size() < 4) {
```
I suggest we use a macro to set the value of the FEW, not just type "4" here

tastynoob replied:
Yes

seanwang2046 commented on src/cpu/o3/issue_queue.cc:
```diff
@@ -623,6 +623,22 @@ Scheduler::SpecWakeupCompletion::description() const
+        "Cycles with no uops executed and at least X in-flight load that is not completed yet"),
+    ADD_STAT(memstall_any_store,
+        "Cycles with few uops executed and no more stores can be issued"),
+    ADD_STAT(memstall_l1miss,
```
the comment here need also modify to "least X"

seanwang2046 commented on src/cpu/o3/lsq.cc:
```diff
@@ -1352,6 +1383,12 @@ LSQ::SingleDataRequest::recvTimingResp(PacketPtr pkt)
     // Dump inst num, request addr, and packet addr
     DPRINTF(LSQ, "Single Req::recvTimingResp: inst: %llu, pkt: %#lx\n", pkt->req->getReqInstSeqNum(),
             pkt->getAddr());
+    if (isLoad()) {
```
what does this part of code do here?

tastynoob replied:
This is used for release the infilght load queue' entry

tastynoob committed:
cpu-o3: add recovery bubble counter and add description

Change-Id: I6388a35015e112c2361b0565ebc3e3278381b476

tastynoob committed:
cpu-o3: ignore fetch bubble if squashing

Change-Id: I7b569cd06c2efbb77ff8f0fa0c50ca995b427dd7

jensen-yan committed:
cpu-o3: add commit total squash counters

Note that: squashDueToBranch, squashDueToOrderViolation are large,
Trap, TC, after are close to zero.

Change-Id: Ic66038976266dbd2202bf8ead18fc7c50675aa24

jensen-yan committed:
cpu-o3: util: add topdown stats at cpu top

merge seperate stats together,
move other stats to the right place,
like iq_io

happy-lx commented on src/cpu/o3/lsq_unit.hh:
```diff
@@ -289,6 +289,8 @@ class LSQUnit
     using LoadQueue = CircularQueue<LQEntry>;
     using StoreQueue = CircularQueue<SQEntry>;
 
+    std::vector<LSQRequest*> inflightLoads;
```
After [#265](https://github.com/OpenXiangShan/GEM5/pull/265) , once an load instruction is replayed due to a cache miss, the request at the corresponding position in the LQ will be set to as `nullptr`, and this miss request will not be tracked. Using `request` to determine if an instruction is inflight might lead to errors.

Maybe use `seqNum` to track inflightLoads can solve these errors

tastynoob committed:
cpu-o3: fix inflightloads squashing

Change-Id: I3ce45d847590a34e54f50ae82347cc5a64b19540
